compiler/ast.nim

```diff
@@ -857,6 +857,7 @@ type
     of routineKinds:
       #procInstCache*: seq[PInstantiation]
       gcUnsafetyReason*: PSym  # for better error messages wrt gcsafe
+      sideEffectReasons*: seq[(TLineInfo, PSym)]  # for better side effect error messages
```

Comment:
instead of changing AST for something that's only used for error msgs, add a Table/HashSet/other to `TContext` (similar to unusedImports or importModuleMap etc, indexed via `PSym.id`)